<div class="container" fxLayout="column" fxLayoutGap="20px">
  <div class="card" style="padding: 0">
    <div class="card-header" style="background-color:#d1e5f0;font-size: 17px">
      Welcome to EndoQuad
    </div>
    <div class="card-body" style="text-align: justify">
      <div class="row">
        <div class="col-8" >
            <p class="big" style="font-size: 16px">G-quadruplexes (G4s) are non-canonical four-stranded structures
              and emerge as novel genetic regulatory elements. However, a comprehensive genomic annotation
              of endogenous G4s (eG4s) and systematic characterization of their regulatory network are
              still lacking, raising great challenges for eG4 research. Here, we present
              EndoQuad to address these pressing issues by integrating
              high-throughput experimental data. First, based on more than 1200 high-quality genome-wide
              eG4s mapping datasets generated by G4 ChIP-seq/CUT&Tag, we generate a reference set of
              genome-wide eG4s across human, mouse, and chicken. Our multi-omics analyses show that most
              eG4s are identified in one or a few cell types. The eG4s with higher occurrences across
              samples are more structurally stable, evolutionarily conserved, enriched in promoter regions,
              mark highly expressed genes, and associate with complex regulatory programs, demonstrating
              higher confidence level for further functional study. Finally, we integrate millions of
              functional genomic variants and prioritize eG4s with regulatory functions in disease
              and cancer context. These efforts have culminated in the comprehensive and
              interactive database of experimentally validated DNA eG4s. As such, EndoQuad
              enables users to easily access, download, and repurpose these data for their own
              research. EndoQuad will become a one-stop resource for eG4 research and lays the
              foundation for future functional study.</p>
          <strong style="margin-bottom: 0px; font-size: 16px">Citation:</strong>
          <p class="big">Qian SH, Shi MW, Xiong YL, et al. EndoQuad: a comprehensive genome-wide experimentally validated endogenous G-quadruplex database. <i>Nucleic Acids Research</i>, 2023: gkad966, <a href="https://doi.org/10.1093/nar/gkad966" target="_blank">https://doi.org/10.1093/nar/gkad966</a>.</p>
        </div>

        <div class="col-4" fxFlexAlign="center">
          <img style="margin-top: -15px;width: 80%" [src]="assets + '/imgs/regulation.png'" class="d-block img-fluid">
        </div>
      </div>
    </div>
  </div>
<!--  <div class="jumbotron" style="margin-top: 10px;padding-bottom: 20px;padding-top: 20px">-->
<!--    <div fxLayout="row" fxLayoutGap="40px">-->
<!--      <div>-->
<!--        <h2>Welcome to EndoQuad</h2>-->
<!--        <div>-->
<!--            <p style="font-size: 16px">G-quadruplexes (G4s) are non-canonical four-stranded structures and-->
<!--              participate in several key cellular processes. However,-->
<!--              a comprehensive reference set and systematical characterization of endogenous G4s-->
<!--              (eG4s) are still lacking. Here we combine the sensitive G4 CUT&Tag method with all-->
<!--              public G4 ChIP-seq data, and exhaustively detect 405,153, 64,116, and 45,584 eG4s in human,-->
<!--              mouse, and chicken, respectively. Native eG4s are prevalent across mammalian genomes and over-->
<!--              83% of human protein-coding genes contain at least one eG4. eG4s are largely dynamic across-->
<!--              diverse cell types, evolutionarily conserved, enriched in promoter regions, and mark highly-->
<!--              expressed genes. We further perform sample-matched ATAC-seq and integrate published epigenomics-->
<!--              data to generate a quantitative eG4 regulome. Constitutively formed eG4s are high-occupancy-->
<!--              targets of transcription factors and substantially contribute to active accessible chromatin-->
<!--              signals, suggesting that they might be novel cis-regulatory elements. To help the researchers-->
<!--              easily browse and access the data, we develop a comprehensive and interactive EndoQuad web-->
<!--              resource. In this database, the users can query the histone modification of eG4s, the interplay-->
<!--              between eG4s and Transcription factors, and colocolization of eG4s with disease/cancer-->
<!--              associated SNPs/eQTLs. The EndoQuad data and resulting principle provided here could-->
<!--              serve as a valuable resource for understanding the function of eG4s and further exploring-->
<!--              their regulatory mechanisms and advancing the identification of therapeutic targets.</p>-->
<!--          <br/><br/>-->
<!--&lt;!&ndash;          <p style="font-size: 16px"><strong style="font-size: 18px">Citation:</strong> EndoQuad: a comprehensive database for endogenous G-quadruplexex across vertebrates</p>&ndash;&gt;-->
<!--        </div>-->
<!--      </div>-->
<!--      <div >-->
<!--        <img style="width: 300px" [src]="assets + '/imgs/regulation.png'" />-->
<!--      </div>-->
<!--    </div>-->
<!--  </div>-->
  <mat-card appMaterialElevation defaultElevation="4" raisedElevation="4">
    <mat-card-header>
      <mat-card-title>Overview of EndoQuad</mat-card-title>
    </mat-card-header>
    <mat-divider inset></mat-divider>
    <div style="margin:10px 80px 0px 80px">
      <img  [src]="assets + '/imgs/functions.png'" usemap="#functions" />
      <map name="functions" id="functions">
        <area shape="rect" coords="0,0,310,190" routerLink="/group-g4" alt="">
        <area shape="rect" coords="407,0,680,190" routerLink="/epigenetics" alt="">
        <area shape="rect" coords="770,0,1078,190" routerLink="/tf-g4" alt="">
        <area shape="rect" coords="103,340,376,542" routerLink="/eqtl-g4" alt="">
        <area shape="rect" coords="680,340,953,542" routerLink="/predicted-g4" alt="">
      </map>
    </div>
  </mat-card>
</div>

